### DISCUSSION TITLE
Can I use a requirements.in file instead of pyproject.toml?

### DISCUSSION THREAD

**Discussion by juftin** [pdm-project/pdm#2392] [Q&A]
I'm looking to leverage PDM for a [hatch](https://github.com/pypa/hatch) LockFile plugin - the current workflow I'm after is using hatch to create a `requirements.in` file - ultimately it's a loosely pinned [requirements file format](https://pip.pypa.io/en/stable/reference/requirements-file-format/#requirements-file-format). I would then feed that file to PDM and create an exhaustive lock file from all the transitive dependencies. I'd really like to leverage PDM for generating the LockFile but it seems like it's only able to infer requirements from a Python build file like a pyproject.toml? 

It's worth noting that the goal here is to create a lock-file for a hatch environment - an example environment might be only used for something like linting and wouldn't include runtime dependencies (although runtime lock files will be another use-case).

**Comment by frostming** [comment]
```
pdm import requirements.in
```

**Comment by juftin** [comment]
> ```
> pdm import requirements.in
> ```

Thanks for the reply, unfortunately [pdm import](https://pdm-project.org/latest/usage/project/#import-the-project-from-other-package-managers) doesn't actually work for my needs.

Ideally I'm looking for something like `pdm lock requirements.in --output locks/lint.yaml` to build a lockfile without having to create a PDM table in my project.toml.

Ultimately the goal is to create a hatch plugin with pdm's locking superpowers.

**Reply by pawamoy** [reply]
A naive solution would be to create a temporary folder, write a pyproject.toml there with your dependencies, run `pdm lock` in that folder to generate the lock file, and copy it to its final destination (`lint.toml`).

Maybe there's a way to use PDM's API to do that without a temporary project.

**Reply by juftin** [reply]
> A naive solution would be to create a temporary folder, write a pyproject.toml there with your dependencies, run `pdm lock` in that folder to generate the lock file, and copy it to its final destination (`lint.toml`).

Yeah, this is a good idea and probably the path I'll go down. I understand PDM is more than just a LockFile generator - so maybe what I'm trying to do is a bit against the spirit of the tool. But this is one feature that `pip-compile` supports that I think wouldn't be terribly hard to do with PDM. FWIW PDM supports almost all the same features I enjoy about hatch so maybe it's time to look at switching over rather than writing a plugin.

**Reply by pawamoy** [reply]
Maybe :smile: 

I think you could use PDM's API without too many difficulties: https://github.com/pdm-project/pdm/blob/main/src/pdm/cli/actions.py#L38.

**Reply by juftin** [reply]
Nice - I will take a look, thank you. That was one of my challenges with trying to use `pip-compile` for this - their API is tangled up with their CLI implementation. I love that `pdm` exposes final API methods like `do_lock` and `do_sync`

**Reply by frostming** [reply]
but be careful about that, PDM doesn't have a clear strategy of the stability of its APIs.

**Reply by pawamoy** [reply]
> their API is tangled up with their CLI implementation

Amen. This is my nemesis. Let me guess... they use Click? Yep.
